### ISSUE TITLE
Hashing nested dicts fails

### ISSUE THREAD

**Issue by maksimt** [spotify/luigi#1898]
Steps to recreate:

``` python
import luigi
class TestClass(luigi.Task):
    par = luigi.DictParameter()

a = TestClass({'a':2,'b':{'c':'3'}})
a.par.__hash__()
```

Expected Error:

``` python
TypeError                                 Traceback (most recent call last)
<ipython-input-1-755c51304007> in <module>()
      4 
      5 a = TestClass({'a':2,'b':{'c':'3'}})
----> 6 a.par.__hash__()

/Users/maks/anaconda2/lib/python2.7/site-packages/luigi/parameter.pyc in __hash__(self)
    790     def __hash__(self):
    791         if self.__hash is None:
--> 792             hashes = map(hash, self.items())
    793             self.__hash = functools.reduce(operator.xor, hashes, 0)
    794 

TypeError: unhashable type: 'dict'
```

Would it be possible to replace this `__hash__()` function with a call to `hashlib.sha1(str(self)).hexdigest()` or just call `hash` on `str(self)` (or replace `str` with JSON Encoder if appropriate).

**Comment by Tarrasch** [comment]
Thanks for your bug report. Actually I'm not sure if this is expected behavior or not (do luigi allow nested dicts??). Can you test if you can run your class from the command line passing a nested dict?

    luigi TestClass --par "{'a':2,'b':{'c':'3'}}"

Perhaps there should be a `PythonValueParameter` that can be any combination of literals, `[]`, `{}` etc., although perhaps it's a bad practice to have it be so untyped.

**Comment by erikbern** [comment]
agree with @Tarrasch

**Comment by maksimt** [comment]
Hi, running it on command line seems to have worked:

```$ luigi --module luigi_dict_test TestClass --par "{'a':2,'b':{'c':'3'}}"```

```
DEBUG: Checking if TestClass(par={'a':2,'b':{'c':'3'}}) is complete
/Users/maks/anaconda2/lib/python2.7/site-packages/luigi/worker.py:305: UserWarning: Task TestClass(par={'a':2,'b':{'c':'3'}}) without outputs has no custom complete() method
  is_complete = task.complete()
INFO: Informed scheduler that task   TestClass___a__2__b____c___2063a9ec8a   has status   PENDING
INFO: Done scheduling tasks
INFO: Running Worker with 1 processes
DEBUG: Asking scheduler for work...
DEBUG: Pending tasks: 1
INFO: [pid 74197] Worker Worker(salt=606770060, workers=1, host=localhost, username=maks, pid=74197) running   TestClass(par={'a':2,'b':{'c':'3'}})
INFO: [pid 74197] Worker Worker(salt=606770060, workers=1, host=localhost, username=maks, pid=74197) done      TestClass(par={'a':2,'b':{'c':'3'}})
DEBUG: 1 running tasks, waiting for next task to finish
INFO: Informed scheduler that task   TestClass___a__2__b____c___2063a9ec8a   has status   DONE
DEBUG: Asking scheduler for work...
DEBUG: Done
DEBUG: There are no more tasks to run at this time
INFO: Worker Worker(salt=606770060, workers=1, host=maksim.cs.rpi.edu, username=maks, pid=74197) was stopped. Shutting down Keep-Alive thread
INFO: 
===== Luigi Execution Summary =====

Scheduled 1 tasks of which:
* 1 ran successfully:
    - 1 TestClass(par={'a':2,'b':{'c':'3'}})

This progress looks :) because there were no failed tasks or missing external dependencies

===== Luigi Execution Summary =====
```

The reason I'm interested in the `__hash__` function is so that I can use it encode the current Task's DictParameter into a string and use it to define Targets in the `output()` and `requires()` methods. If there's a existing workflow that achieves this with DictParameters, I would appreciate a pointer to an example.

**Comment by Tarrasch** [comment]
@maksimt. I can't tell from the [docs](http://luigi.readthedocs.io/en/stable/api/luigi.parameter.html#luigi.parameter.DictParameter) whether if we ever intended users to be able to have nested dicts. I suppose the feature could be added. If you are ready to do that. I could give you some guidance for implementing it. It's not too hard I believe.

**Comment by maksimt** [comment]
Hi, yes I'm interested in implementing this feature with some guidance. I'm not sure if the FrozenDict's `__hash__()` method is the only thing that needs to be changed, since in other parts of the code hashability is checked without using this method. For example the debug statement I see in my use case comes from `task_register.py`:

```python
        k = (cls, tuple(param_values))
        try:
            hash(k)
        except TypeError:
            logger.debug("Not all parameter values are hashable so instance isn't coming from the cache")
            return instantiate()  # unhashable types in parameters
```

However, after hashing the nested dict as follows in order to define `Target`s in `output()`, luigi realizes that it doesn't have to recompute those targets, and in fact retrieves the previously computed targets.

```python

def format_param_to_file_system(_DictParameter):
    """formats string so it can be used as a file name"""
    b = luigi.DictParameter()
    str = b.serialize(_DictParameter)
    return hashlib.sha1(str).hexdigest()

def output(self):
     return HdfsTarget('{path}/local_models|{id}.pickle'.format(
                        path=self.settings['exp_path'],
                        id=format_param_to_file_system(self.settings)
                         )
                       )
```

So I'm not sure exactly what parts of the codebase should be modified. I've also been using luigi for less than a week, so I'm not sure if my issue is simply wrong workflow vs lack of a feature.

**Comment by Tarrasch** [comment]
@maksimt, We don't need to with making all parameters immutable+hashable. Lets do it one by one. Why don't you start with DictParameter?

**Comment by maksimt** [comment]
I believe this issue is fixed in 2.7.1. I was about to dig in, but the code from the opening post of this issue no longer raises an error. Rather it returns a hash, `5651538082574827812`.
